### ISSUE TITLE
Improve parameter handling in get_ontotrace_data

### ISSUE THREAD

**Issue by johnbradley** [phenoscape/rphenoscape#180]
Improve parameter handling in [get_ontotrace_data()](https://github.com/phenoscape/rphenoscape/blob/ff42f3f355fc87c9293aecacbf520f1b5bd6b073/R/pk_ontotrace.R#L120-L174).

This issue was broken out of PR#176.
Discussion from that PR: https://github.com/phenoscape/rphenoscape/pull/176#issuecomment-785123476

**Comment by johnbradley** [comment]
Summary of discussion about `get_ontotrace_data()` parameter handling from PR #176:

`get_ontotrace_data()` currently creates OWL expressions for the `taxon` and `entity` function parameters.
Creating an OWL expression for `taxon` is not necessary, but for the `entity` field an OWL expression is required to support the `relation` function parameter. The `relation` parameter can be either "part of" or "develops from".
These OWL expressions are used with the `/ontotrace` endpoint. The `/ontotrace` endpoint allows passing an OWL expression and/or IRI lists using the _taxon_, _taxon_list_, _entity_ and _entity_list_ API parameters. For _taxon_list_ and  _entity_list_ API params only those terms are returned as rows or columns - No subsumption query happens at all.

---

Additional notes from @hlapp 

> The documentation for the `entity` function parameter says a user can pass `anatomical class expressions`:
https://github.com/phenoscape/rphenoscape/blob/f584052601a37d5b0d4876c7082d196d4e2b5712/R/pk_get_xml.R#L14-L15
> This is incorrect. The function assumes this parameter contains a vector of terms or IRI.

**Comment by johnbradley** [comment]
The [parameters for get_ontotrace_data()](https://github.com/phenoscape/rphenoscape/blob/67e06ccd0f954f3feb0487681ee813adc1b7e381/R/pk_ontotrace.R#L74-L89) and documentation are as follows:
- `taxon` character, required
     - >a vector of taxon names
- `entity` character, required
    - > A single character string or a vector of anatomical class expressions
- `relation` character string, optional (default "part of")
    - >The relationship to the entities to be included in the result. Must be
either "part of" or "develops from", or set to NA to disable.
Default is "part of".
- `variable_only` logical, optional
   - >Whether to only include characters that are variable across the selected set of taxa. Default is TRUE.
- `strict` logical, optional
    - >Whether or not to treat any failure to resolve
any taxon or entity names to IRI as input error. Resolution by partial or
other inexact match results in a warning, but is not considered a failure.
If FALSE, query execution will continue with the taxon and entity terms
that did resolve to IRI. Default is TRUE, meaning any resolution failure will
result in an error.

__NOTE:__ The documentation for `entity` is wrong. Based on the code it would be something like `A vector of anatomical names`. Also there is an undocumented feature that `taxon` and `entity` will also accept a vector of IRIs.

---

The `/ontotrace` endpoint parameters are as follows
- `taxon` 
   - >Taxonomic class expression in OWL Manchester syntax
- `taxon_list`
   - >JSON array of term IRIs
- `entity`
   - >Anatomical class expression in OWL Manchester syntax
- `entity_list`
   - >JSON array of term IRIs
- `variable_only`
   - >whether to restrict the matrix to characters with both ‘present’ and ‘absent’ values in the matrix

__NOTE:__ Only a single class expression is allowed as a parameter for `taxon` or `entity`.

**Comment by johnbradley** [comment]
My plan is to introduce an `Owl` class that contains a "class expression in OWL Manchester syntax". Then the `taxon` and `entity` parameters of get_ontotrace_data() would support a vector of names, a vector of IRI or a single class expression. One issue with this plan is the `relation` parameter( with default "part of") since it would only work with a vector of names/IRI. I could remove the `relation` parameter and add a new function to create an `Owl` instance that combines the relation with a vector of names/IRIs. Thoughts?

**Comment by hlapp** [comment]
Couldn't we simply `stop()` if `relation` is provided when an OWL class expression is also provided? Or document that the parameter would be ignored in this situation? (Neither option would be unusual in R.)

Also, I think we'll want an additional parameter to indicate that no reasoning is to be used for taxon and entities (meaning, this would be passed to the API's `taxon_list` and `entity_list` parameters), i.e., the taxa and entities for which data are returned will be _only_ those in the provided list, not ones subsumed by them. (In this case, the `relation` parameter would also necessarily have to be ignored, since it wouldn't make any sense if no reasoning were requested.)

**Comment by johnbradley** [comment]
> Couldn't we simply stop() if relation is provided when an OWL class expression is also provided? Or document that the parameter would be ignored in this situation? (Neither option would be unusual in R.)

Since relation has a default value of "part of" it would (to me at least) make it confusing to have to pass NULL or NA to disable it. Documenting that the parameter is ignored sounds good. 

> Also, I think we'll want an additional parameter to indicate that no reasoning is to be used for taxon and entities (meaning, this would be passed to the API's taxon_list and entity_list parameters), i.e., the taxa and entities for which data are returned will be only those in the provided list, not ones subsumed by them. (In this case, the relation parameter would also necessarily have to be ignored, since it wouldn't make any sense if no reasoning were requested.)

Isn't this what the `NA` value passed to `relation` is trying to achieve?
https://github.com/phenoscape/rphenoscape/blob/67e06ccd0f954f3feb0487681ee813adc1b7e381/R/pk_ontotrace.R#L77-L80

**Comment by johnbradley** [comment]
I would have never thought you could ask a variable if it was passed in or was a default value. 🤯

**Comment by johnbradley** [comment]
@balhoff I am trying to use the `taxon_list` parameter with the [kb.phenoscape /ontotrace endpoint](https://kb.phenoscape.org/apidocs/#/OntoTrace/get_ontotrace) and getting mostly empty results. Since the Swagger UI doesn't seem to support JSON arrays the example below is using curl.
I am passing the taxon_list param as a URL encoded value of `["http://purl.obolibrary.org/obo/VTO_0036217"]`.
I am passing the entity param as a URL encoded value of `(<http://purl.obolibrary.org/obo/UBERON_0008897> or <http://purl.obolibrary.org/obo/BFO_0000050>  some  <http://purl.obolibrary.org/obo/UBERON_0008897>)`

The curl command:
```
curl "https://kb.phenoscape.org/api/ontotrace?taxon_list=%5B%22http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FVTO_0036217%22%5D&entity=%28%3Chttp%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FUBERON_0008897%3E%20or%20%3Chttp%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FBFO_0000050%3E%20%20some%20%20%3Chttp%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FUBERON_0008897%3E%29&variable_only=TRUE"
```
The mostly empty response:
```
<?xml version="1.0" encoding="UTF-8"?>
<nexml xmlns="http://www.nexml.org/2009" xmlns:dc="http://purl.org/dc/terms/" xmlns:dwc="http://rs.tdwg.org/dwc/terms/" xmlns:obo="http://purl.obolibrary.org/obo/" xmlns:ps="http://vocab.phenoscape.org/" xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="0.9" xsi:schemaLocation="http://www.nexml.org/2009 http://www.nexml.org/2009/nexml.xsd http://www.bioontologies.org/obd/schema/pheno http://purl.org/phenoscape/phenoxml.xsd">
  <meta xsi:type="LiteralMeta" property="dc:creator">Phenoscape KB built on 2020-04-08T20:37:21.892Z</meta>
  <meta xsi:type="LiteralMeta" property="dc:description">Generated from the Phenoscape Knowledgebase on 2021-4-30 by Ontotrace query :
* taxa: Not provided
* taxa list: [http://purl.obolibrary.org/obo/VTO_0036217] 
* entities: &lt;http://purl.obolibrary.org/obo/UBERON_0008897&gt; or (&lt;http://purl.obolibrary.org/obo/BFO_0000050&gt; some &lt;http://purl.obolibrary.org/obo/UBERON_0008897&gt;)
* entity list: Not provided 
* variable only: true</meta>
  <otus id="tf94cadfe-ad26-40db-8b4b-14b2f3a484ac" />
  <characters id="cb9119978-04e9-472a-ac3d-4883d84f4cec" xsi:type="StandardCells" otus="tf94cadfe-ad26-40db-8b4b-14b2f3a484ac">
    <format />
    <matrix />
  </characters>
  <trees id="tc84ed4f9-96f1-4cb9-962f-841769d5f79a" otus="tf94cadfe-ad26-40db-8b4b-14b2f3a484ac" />
</nexml>
```
If I use the `taxon` param with the URL encoded value of `<http://purl.obolibrary.org/obo/VTO_0036217>` I get back data.
The curl command:
```
curl "https://kb.phenoscape.org/api/ontotrace?taxon=%3Chttp%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FVTO_0036217%3E&entity=%28%3Chttp%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FUBERON_0008897%3E%20or%20%3Chttp%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FBFO_0000050%3E%20%20some%20%20%3Chttp%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FUBERON_0008897%3E%29&variable_only=TRUE"
```
The response:
```
<?xml version="1.0" encoding="UTF-8"?>
<nexml xmlns="http://www.nexml.org/2009" xmlns:dc="http://purl.org/dc/terms/" xmlns:dwc="http://rs.tdwg.org/dwc/terms/" xmlns:obo="http://purl.obolibrary.org/obo/" xmlns:ps="http://vocab.phenoscape.org/" xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="0.9" xsi:schemaLocation="http://www.nexml.org/2009 http://www.nexml.org/2009/nexml.xsd http://www.bioontologies.org/obd/schema/pheno http://purl.org/phenoscape/phenoxml.xsd">
  <meta xsi:type="LiteralMeta" property="dc:creator">Phenoscape KB built on 2020-04-08T20:37:21.892Z</meta>
  <meta xsi:type="LiteralMeta" property="dc:description">Generated from the Phenoscape Knowledgebase on 2021-4-30 by Ontotrace query :
* taxa: &lt;http://purl.obolibrary.org/obo/VTO_0036217&gt;
* taxa list: Not provided 
* entities: &lt;http://purl.obolibrary.org/obo/UBERON_0008897&gt; or (&lt;http://purl.obolibrary.org/obo/BFO_0000050&gt; some &lt;http://purl.obolibrary.org/obo/UBERON_0008897&gt;)
* entity list: Not provided 
* variable only: true</meta>
  <otus id="t31ec52e5-535c-440c-81b7-7ecfe5fdeeb5">
    <otu id="VTO_0061498" label="Ictalurus mexicanus" about="#VTO_0061498">
      <meta xsi:type="ResourceMeta" rel="dwc:taxonID" href="http://purl.obolibrary.org/obo/VTO_0061498" />
    </otu>
    <otu id="VTO_0036220" label="Ictalurus lupus" about="#VTO_0036220">
      <meta xsi:type="ResourceMeta" rel="dwc:taxonID" href="http://purl.obolibrary.org/obo/VTO_0036220" />
    </otu>
...
```
Any idea what I am doing wrong in the `taxon_list` curl example? Thanks.

**Comment by hlapp** [comment]
@johnbradley did you try setting `variable_only` to FALSE? If there's only one taxon then by definition there cannot be variation between taxa, and so without `variable_only = FALSE` you'll necessarily get empty results back.

**Comment by hlapp** [comment]
> If there's only one taxon then by definition there cannot be variation between taxa

Actually I take that back, with subsumption there certainly could. But the argument still holds, that there may not be character variation at this taxon.

**Comment by johnbradley** [comment]
@hlapp If I change `variable_only` to FALSE I still get back empty nexml object:
```
$ curl "https://kb.phenoscape.org/api/ontotrace?taxon_list=%5B%22http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FVTO_0036217%22%5D&entity=%28%3Chttp%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FUBERON_0008897%3E%20or%20%3Chttp%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FBFO_0000050%3E%20%20some%20%20%3Chttp%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FUBERON_0008897%3E%29&variable_only=FALSE"
<?xml version="1.0" encoding="UTF-8"?>
<nexml xmlns="http://www.nexml.org/2009" xmlns:dc="http://purl.org/dc/terms/" xmlns:dwc="http://rs.tdwg.org/dwc/terms/" xmlns:obo="http://purl.obolibrary.org/obo/" xmlns:ps="http://vocab.phenoscape.org/" xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="0.9" xsi:schemaLocation="http://www.nexml.org/2009 http://www.nexml.org/2009/nexml.xsd http://www.bioontologies.org/obd/schema/pheno http://purl.org/phenoscape/phenoxml.xsd">
  <meta xsi:type="LiteralMeta" property="dc:creator">Phenoscape KB built on 2020-04-08T20:37:21.892Z</meta>
  <meta xsi:type="LiteralMeta" property="dc:description">Generated from the Phenoscape Knowledgebase on 2021-4-30 by Ontotrace query :
* taxa: Not provided
* taxa list: [http://purl.obolibrary.org/obo/VTO_0036217] 
* entities: &lt;http://purl.obolibrary.org/obo/UBERON_0008897&gt; or (&lt;http://purl.obolibrary.org/obo/BFO_0000050&gt; some &lt;http://purl.obolibrary.org/obo/UBERON_0008897&gt;)
* entity list: Not provided 
* variable only: false</meta>
  <otus id="t01b1ea30-b4d0-4da3-815a-54605e7197a5" />
  <characters id="c4f8ecaca-1f91-4fe9-8e37-c71dc16dd6a7" xsi:type="StandardCells" otus="t01b1ea30-b4d0-4da3-815a-54605e7197a5">
    <format />
    <matrix />
  </characters>
  <trees id="t30a509c6-b736-40b7-9065-35642dcde67d" otus="t01b1ea30-b4d0-4da3-815a-54605e7197a5" />
</nexml>
```

Perhaps this is a misunderstanding on my part. I thought that passing `<http://purl.obolibrary.org/obo/VTO_0036217>` to the `taxon` parameter would be equivalent to passing`["http://purl.obolibrary.org/obo/VTO_0036217"]` to the `taxon_list` parameter.

**Comment by balhoff** [comment]
There is no subsumption query for the taxa in the `taxon_list`. So maybe try using one of the taxon ids that you get in the results when you put that taxon in the `taxon` parameter.

**Comment by hlapp** [comment]
> I thought that passing `<http://purl.obolibrary.org/obo/VTO_0036217>` to the `taxon` parameter would be equivalent to passing`["http://purl.obolibrary.org/obo/VTO_0036217"]` to the `taxon_list` parameter.

No it's not (even though in this particular case it actually is). As I said, the latter doesn't do subsumption, so you only get data for exactly the taxa you pass, not also for the taxa that are subclasses. In your particular example, this is a genus, and it so happens that the lowest taxonomic rank at which character variation is in the database is genus. So in this case, including subclass taxa (i.e., the Ictalurus species) won't surface any more variable (or non-variable) Ontotrace data.

**Comment by johnbradley** [comment]
Thanks @balhoff and @hlapp. I think I understand now.

**Comment by johnbradley** [comment]
Fixed by #197
